pysph/base/tree/helpers.py

```diff
@@ -0,0 +1,147 @@
+import numpy as np
+import pyopencl as cl
+import pyopencl.array
+from pyopencl.elementwise import ElementwiseKernel
+from pytools import memoize
+from pysph.base.opencl import get_context, get_queue
+
+_cache = {}
+
+
+@memoize
+def get_copy_kernel(ctx, dtype1, dtype2):
+    return ElementwiseKernel(
+        ctx,
+        """
+        %(data_t1)s *x1, %(data_t1)s *y1, %(data_t1)s *z1, %(data_t1)s *h1,
+        %(data_t2)s *x2, %(data_t2)s *y2, %(data_t2)s *z2, %(data_t2)s *h2
+        """ % dict(
+            data_t1=dtype1,
+            data_t2=dtype2
+        ),
+        operation="""
+        x2[i] = (%(data_t2)s)x1[i];
+        y2[i] = (%(data_t2)s)y1[i];
+        z2[i] = (%(data_t2)s)z1[i];
+        h2[i] = (%(data_t2)s)h1[i];
+        """ % dict(
+            data_t2=dtype2
+        )
+    )
+
+
+def cache_result(cache_key):
```

Comment:
This doesn't seem to be used at all and can be removed?

Reply:
It was being used previously but I've removed it now.